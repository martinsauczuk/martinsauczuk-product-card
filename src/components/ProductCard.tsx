import React, { CSSProperties, createContext } from 'react';
import { useProduct } from '../hooks/useProduct';
import { Product, ProductContextProps, OnChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;


export interface Props {
    // children?: ReactElement | ReactElement[];
    children: ( args: ProductCardHandlers ) => JSX.Element;
    product: Product;
    className?: string;
    style?: CSSProperties;
    onChange?: ( args: OnChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues
}

export const ProductCard = ({children, product, className, style, onChange, value, initialValues }: Props ) => {

    const { increaseBy, counter, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

    return (

        <Provider value = {{
            counter,
            maxCount,
            increaseBy,
            product,
        }}>
            <div 
                className={ `${styles.productCard} ${className}` }
                style={style}
            >
                { 
                    children({
                        count: counter,
                        isMaxCountReached,
                        maxCount: initialValues?.maxCount,
                        product,
                        increaseBy,
                        reset,
                    })
                }
            </div>
        </Provider>
    )
}
