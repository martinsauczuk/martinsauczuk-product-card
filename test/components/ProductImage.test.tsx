import React from "react";
import ProductCard, { ProductImage } from "../../src/components";
import { product2 } from "../data/products";
import renderer from 'react-test-renderer';

describe('ProductImage', () => {

    test('should show the component with image', () => {
      const wrapper = renderer.create(
          <ProductImage img="https://hola.jpg" />
      )
      expect( wrapper.toJSON() ).toMatchSnapshot();
    });

  
    test('should show the componente with the image of the product', () => {
      
      const wrapper = renderer.create(
          <ProductCard product={ product2 } >
              {
                  () => <ProductImage />
              }
          </ProductCard>
      )
      expect( wrapper.toJSON() ).toMatchSnapshot();
  
    });
    
  })