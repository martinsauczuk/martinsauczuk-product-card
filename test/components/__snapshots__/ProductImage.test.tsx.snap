// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should show the component with image 1`] = `
<img
  alt="Product image"
  className="productImg undefined"
  src="https://hola.jpg"
/>
`;

exports[`ProductImage should show the componente with the image of the product 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product image"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
